### PR TITLE
Schema and read endpoints Pool

### LINKED ISSUE #8131: API Endpoints - Read-only - Pools
**Description**
Hello 

We need to create several endpoints that perform basic read-only operations on **Pools** . We need the following endpoints:

- GET /pools
- GET /pools/{pool_name}

Detailed information is available in the issue:
https://github.com/apache/airflow/issues/8118

**Use case / motivation**
N/A

**Related Issues**
N/A

### PR DESCRIPTION
**Pull request by OmairK**
Closes #8131 
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/api_connexion/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+def response_404(
```
Comment (mik-laj) [inline airflow/api_connexion/exceptions.py]: I think, we should use exception to handle errors. 
I have prepared an example that shows how to do it.
https://termbin.com/legq
To add this commit to your brach, run the following command:
```
curl https://termbin.com/legq | git am
```

Comment (OmairK) [reply airflow/api_connexion/exceptions.py]: Thank you this is exactly what I was stuck on.

Comment (OmairK) [reply airflow/api_connexion/exceptions.py]: Fixed `00da53d `

Comment (OmairK) [reply airflow/api_connexion/exceptions.py]: Added other http exception classes `cd126b7`

File: airflow/api_connexion/endpoints/pool_endpoint.py [hunk 2]
```diff
@@ -26,18 +30,33 @@ def delete_pool():
...
 
-def get_pools():
+@provide_session
+def get_pools(session):
     """
     Get all pools
     """
-    raise NotImplementedError("Not implemented yet.")
+    offset = request.args.get(parameters.page_offset, 0)
+    limit = min(request.args.get(parameters.page_limit, 100), 100)
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/pool_endpoint.py]:
```suggestion
    limit = min(int(request.args.get(parameters.page_limit, 100)), 100)
```

Comment (ephraimbuddy) [reply airflow/api_connexion/endpoints/pool_endpoint.py]: You can add tests to handle pagination too

Comment (OmairK) [reply airflow/api_connexion/endpoints/pool_endpoint.py]: Thanks
Here is the change `983e673`

File: airflow/api_connexion/endpoints/pool_endpoint.py [hunk 3]
```diff
@@ -26,18 +30,33 @@ def delete_pool():
...
     """
-    raise NotImplementedError("Not implemented yet.")
+    offset = request.args.get(parameters.page_offset, 0)
+    limit = min(int(request.args.get(parameters.page_limit, 100)), 100)
+
+    query = session.query(Pool)
+    query = query.offset(offset).limit(limit)
+
+    pools = query.all()
+    return pool_collection_schema.dump(PoolCollection(pools=pools, total_entries=len(pools))).data
```
Comment (mik-laj) [inline airflow/api_connexion/endpoints/pool_endpoint.py]: Total entries should contain the number of all items, not just the number of items on the current page.

File: tests/api_connexion/endpoints/test_pool_endpoint.py [hunk 4]
```diff
@@ -16,51 +16,143 @@
...
+    def test_limit(self, session):
+        pools = [Pool(pool=f"test_pool{i}", slots=1) for i in range(120)]
+        session.add_all(pools)
+        session.commit()
+        result = session.query(Pool).count()
+        self.assertEqual(result, 121)  # accounts for default pool as well
+        # Test for limit under 100
+        response = self.client.get("/api/v1/pools?limit=10")
+        assert response.status_code == 200
+        self.assertEqual(response.json.get("total_entries"), 10)
```
Comment (mik-laj) [inline tests/api_connexion/endpoints/test_pool_endpoint.py]:
```suggestion
        self.assertEqual(response.json["total_entries"], 121)
```
It should return count of all items.

File: tests/api_connexion/endpoints/test_pool_endpoint.py [hunk 5]
```diff
@@ -16,51 +16,143 @@
...
+        # Test for limit over 100
+        response = self.client.get("/api/v1/pools?limit=110")
         assert response.status_code == 200
+        self.assertEqual(response.json.get('total_entries'), 100)
```
Comment (mik-laj) [inline tests/api_connexion/endpoints/test_pool_endpoint.py]:
```suggestion
        self.assertEqual(response.json['total_entries'], 121)
```
It should return the count of all items.

File: airflow/api_connexion/endpoints/pool_endpoint.py (hunk 3 above)
Comment (OmairK) [reply airflow/api_connexion/endpoints/pool_endpoint.py]: Thanks.
Here is the change ~ce9b79f~  36d3a3f

File: tests/api_connexion/endpoints/test_pool_endpoint.py (hunk 4 above)
Comment (OmairK) [reply tests/api_connexion/endpoints/test_pool_endpoint.py]: Here is the change ~ce9b79f~  36d3a3f

File: tests/api_connexion/endpoints/test_pool_endpoint.py (hunk 5 above)
Comment (OmairK) [reply tests/api_connexion/endpoints/test_pool_endpoint.py]: Here is the change ~ce9b79f~  36d3a3f

File: tests/api_connexion/schemas/test_pool_schemas.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    def tearDown(self) -> None:
+        clear_db_pools()
+
+    def test_serialize(self):
+        pool_model_a = Pool(pool="test_pool_a", slots=3)
+        pool_model_b = Pool(pool="test_pool_b", slots=3)
+        instance = PoolCollection(
+            pools=[pool_model_a, pool_model_b], total_entries=2)
+        self.assertEqual(
+            {"pools": [{"name": "test_pool_a",
```
Comment (mik-laj) [inline tests/api_connexion/schemas/test_pool_schemas.py]: Can you format his file with black? This source code snippet looks weird.

Comment (OmairK) [reply tests/api_connexion/schemas/test_pool_schemas.py]: Sure, I was using autopep8 from the start I will use black from now on.

Comment (OmairK) [reply tests/api_connexion/schemas/test_pool_schemas.py]: Here is the change `98ecf3b`

**Review by mik-laj — APPROVED** [review]
Perfect. I merge it when I get permission from Kaxil.
https://apache-airflow.slack.com/archives/CCPRP7943/p1591962654060300

**Comment by mik-laj** [comment]
Github Action is sad. I think these are a flaky tests. However, to be sure, could you do a rebase?  Over 3 red crosses are not good.

**Comment by OmairK** [comment]
> Github Action is sad. I think these are a flaky tests. However, to be sure, could you do a rebase? Over 3 red crosses are not good.

All the api_connexion tests are passing.
Just one test is failing 

> tests/www/test_views.py

**Comment by mik-laj** [comment]
@OmairK Can you do a rebase? We fixed test_views.py. https://github.com/apache/airflow/pull/9307

**Pull request merged**